src/app-logic/browser-connection.js

```diff
@@ -0,0 +1,179 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+
+import { oneLine } from 'common-tags';
+import {
+  getProfileViaWebChannel,
+  getSymbolTableViaWebChannel,
+  querySupportsGetProfileAndSymbolicationViaWebChannel,
+} from './web-channel';
+
+/**
+ * This file manages the communication between the profiler and the browser.
+ */
+
+export type BrowserConnectionStatus =
+  // In non-Firefox browsers we don't attempt to establish a connection.
+  // This is determined via the userAgent.
+  | {| status: 'NOT_FIREFOX' |}
+  // We are in Firefox, and have sent the initial WebChannel event.
+  | {| status: 'WAITING' |}
+  // We are in Firefox but the WebChannel connection has been denied.
+  // This usually means that this profiler instance is running on a
+  // different host than the one that's specified in the
+  // preference `devtools.performance.recording.ui-base-url`.
+  | {| status: 'DENIED', error: Error |}
+  // We are in Firefox but the WebChannel did not respond within 5 seconds.
+  // This is unexpected. It could mean that we are running in an old Firefox
+  // (older than Firefox 76) which did not have a profiler WebChannel.
+  | {| status: 'TIMED_OUT' |}
+  // The WebChannel connection has been established.
+  | {| status: 'ESTABLISHED', browserConnection: BrowserConnection |};
+
+/**
+ * The interface of communication with the browser. Can be backed by a WebChannel
+ * or by the frame script API.
+ * Only exists if at least an old version of the WebChannel is available in this browser.
+ */
+export interface BrowserConnection {
```

Comment:
Praise: thanks for this more object-oriented code to better encapsulate this behavior.